src/Hedgehog/Internal/Function.idr

```diff
@@ -0,0 +1,351 @@
+        clause : (Nat, Con _ _) -> Clause
+        clause (idx, con) = accumArgs unerased (\x => `(~(var fun) ~x)) rhs arg con where
+
+          arg : BoundArg 1 Unerased -> TTImp
+          arg $ BA g [x] _ = assertIfRec nms g.type `(Function.perturb ~(varStr x) . Function.shiftArg)
+
+          rhs : SnocList TTImp -> TTImp
+          rhs = foldr (\l, r => `(~l . ~r)) `(Seed.variant ~(primVal $ B64 $ cast idx))
+
+      cogenImplClaim : Name -> ParamTypeInfo -> Decl
+      cogenImplClaim impl p = implClaimVis vis impl $ implType "Cogen" p
+
+      cogenImplDef : (fun, impl : Name) -> Decl
+      cogenImplDef fun impl = def impl $ pure $ patClause (var impl) (var "MkCogen" `app` var fun)
+
+
+  ||| Alias for `CogenVis Public`
+  export %inline
+  Cogen : List Name -> ParamTypeInfo -> Res (List TopLevel)
+  Cogen = CogenVis Public
+
+||| Generates a random function being given a generator of codomain type
+|||
+||| This function takes a co-generator of domain type using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+||| If you need a showable function, you have to use a shrinkable version,
+||| which requires more strict implementation on the domain type.
+export
+function_ : Cogen a => Gen b -> Gen (a -> b)
+function_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen bg sz $ perturb x sd
+
+||| Generates a random dependently typed function being given a generator of codomain type family
+|||
+||| This function takes a co-generator of domain type using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+depfun_ : Cogen a => {0 b : a -> Type} -> ((x : a) -> Gen $ b x) -> Gen ((x : a) -> b x)
+depfun_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen (bg x) sz $ perturb x sd
+
+||| Generates a random function with dependently typed domain being given a generator of codomain type
+|||
+||| This function takes a co-generator of domain type family using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+dargfun_ : {0 b : a -> Type} -> ({0 x : a} -> Cogen (b x)) => Gen c -> Gen ({0 x : a} -> b x -> c)
+dargfun_ bg = MkGen $ \sz, sd => singleton $ \x => value $ unGen bg sz $ perturb x sd
+
+||| Generates a random dependently typed function with dependently typed domain
+||| being given a generator of codomain type family
+|||
+||| This function takes a co-generator of domain type family using `auto`-argument based on the type.
+||| This generator does not shrink.
+|||
+||| Notice that this generator returns a non-showable value (unless you invent your own implementation).
+export
+dargdepfun_ : {0 b : a -> Type} ->
```

Comment:
Please use the following style for multi-line function declarations:

```idris
dargdepfun_ : 
     {0 b : a -> Type}
  -> {0 c : {0 x : a} -> b x -> Type}
  -> {auto _ : {0 x : a} -> Cogen (b x)}
  -> ({0 x : a} -> (y : b x) -> Gen (c y))
  -> Gen ({0 x : a} -> (y : b x) -> c y)
```

I try to use this style consistently throughout my codebase nowadays, but have not adjusted older code yet.

Reply:
Sure, I just forgot. Since this was a new module, I used the style I'm used to without any additional thought, since there was no visual reference, as usual when I edit some existing module

Reply:
Looking through the hedgehog code I realize it needs some overhaul to be adjusted to my current coding style (it's one of my oldes Idris projects). I'll do that once this PR is shipped.

Reply:
Done